File: esp32/scanner/main/detection/dji_drone_id_parser.c
/**
 * Friend or Foe — DJI DroneID Parser (ESP32-S3)
 *
 * Ported from Android DjiDroneIdParser.kt.
 * Parses DJI-proprietary vendor-specific IEs from WiFi beacons.
 *
 * IE format (after 3-byte OUI 0x26 0x37 0x12):
 *   Byte 0:      Version/type indicator
 *   Bytes 1-4:   Serial number prefix (first 4 chars, ASCII)
 *   Bytes 5-8:   Drone longitude (int32, degrees x 1e7)
 *   Bytes 9-12:  Drone latitude (int32, degrees x 1e7)
 *   Bytes 13-14: Altitude (int16, meters relative to takeoff)
 *   Bytes 15-16: Height above ground (int16, decimeters)
 *   Bytes 17-18: Speed (uint16, cm/s)
 *   Bytes 19-20: Heading (int16, degrees x 100, 0 = North)
 *   Bytes 21-24: Home longitude (int32, degrees x 1e7)
 *   Bytes 25-28: Home latitude (int32, degrees x 1e7)
 */

#include "dji_drone_id_parser.h"
#include "constants.h"

#include <string.h>
#include <math.h>
#include <esp_log.h>

static const char *TAG = "dji_parser";

/* ── Little-endian read helpers ─────────────────────────────────────────── */

static int32_t read_int32_le(const uint8_t *data, size_t offset, size_t len)
{
    if (offset + 4 > len) return 0;
    return (int32_t)(
        ((uint32_t)data[offset])           |
        ((uint32_t)data[offset + 1] << 8)  |
        ((uint32_t)data[offset + 2] << 16) |
        ((uint32_t)data[offset + 3] << 24)
    );
}

static int16_t read_int16_le(const uint8_t *data, size_t offset, size_t len)
{
    if (offset + 2 > len) return 0;
    uint16_t val = (uint16_t)(
        ((uint16_t)data[offset]) |
        ((uint16_t)data[offset + 1] << 8)
    );
    return (int16_t)val;
}

static uint16_t read_uint16_le(const uint8_t *data, size_t offset, size_t len)
{
    if (offset + 2 > len) return 0;
    return (uint16_t)(
        ((uint16_t)data[offset]) |
        ((uint16_t)data[offset + 1] << 8)
    );
}

/* ── Internal payload parser ────────────────────────────────────────────── */

/**
 * Parse the DroneID payload bytes (after OUI has been stripped).
 * `data` points to the first byte after the 3-byte OUI.
 * `data_len` is the number of bytes available after OUI.
 */
static bool parse_payload(const uint8_t *data, size_t data_len, dji_drone_id_data_t *out)
{
    if (data_len < DJI_MIN_PAYLOAD_LENGTH) {
        ESP_LOGD(TAG, "DroneID IE too short: %zu bytes (need %d)", data_len, DJI_MIN_PAYLOAD_LENGTH);
        return false;
    }

    memset(out, 0, sizeof(*out));

    out->version = data[0];

    /* Drone position: degrees x 1e7 as signed 32-bit integer */
    double drone_lon = (double)read_int32_le(data, 5, data_len) / 1e7;
    double drone_lat = (double)read_int32_le(data, 9, data_len) / 1e7;

    /* Validate coordinates */
    if (drone_lat < -90.0 || drone_lat > 90.0 || drone_lon < -180.0 || drone_lon > 180.0) {
        ESP_LOGD(TAG, "Invalid coordinates in DroneID: lat=%.7f, lon=%.7f", drone_lat, drone_lon);
        return false;
    }
    if (drone_lat == 0.0 && drone_lon == 0.0) {
        ESP_LOGD(TAG, "Zero coordinates in DroneID — GPS not acquired");
        return false;
    }

    out->latitude = drone_lat;
    out->longitude = drone_lon;

    /* Altitude relative to takeoff (int16, meters) */
    out->altitude_m = (double)read_int16_le(data, 13, data_len);

    /* Height above ground (int16, decimeters -> meters) */
    out->height_above_ground_m = (double)read_int16_le(data, 15, data_len) / 10.0;

    /* Speed (uint16, cm/s -> m/s) */
    uint16_t speed_cms = read_uint16_le(data, 17, data_len);
    out->speed_mps = (float)speed_cms / 100.0f;

    /* Heading (int16, degrees x 100 -> degrees) */
    int16_t heading_raw = read_int16_le(data, 19, data_len);
    out->heading_deg = (float)heading_raw / 100.0f;

    /* Home point / operator position */
    double home_lon = (double)read_int32_le(data, 21, data_len) / 1e7;
    double home_lat = (double)read_int32_le(data, 25, data_len) / 1e7;

    if (home_lat != 0.0 || home_lon != 0.0) {
        out->home_lat = home_lat;
        out->home_lon = home_lon;
        out->has_home = true;
    } else {
        out->home_lat = 0.0;
        out->home_lon = 0.0;
        out->has_home = false;
    }

    /* Serial number prefix: ASCII bytes at offset 1-4 */
    size_t serial_copy = 4;
    if (serial_copy > data_len - 1) {
        serial_copy = data_len - 1;
    }
    memcpy(out->serial_prefix, &data[1], serial_copy);
    out->serial_prefix[serial_copy] = '\0';

    /* Trim trailing nulls from serial */
    for (int i = (int)serial_copy - 1; i >= 0; i--) {
        if (out->serial_prefix[i] == '\0' || out->serial_prefix[i] == ' ') {
            out->serial_prefix[i] = '\0';
        } else {
            break;
        }
    }

    ESP_LOGD(TAG, "Parsed DroneID: lat=%.7f, lon=%.7f, alt=%.0fm, "
             "speed=%.1fm/s, heading=%.1f°, serial=%s",
             out->latitude, out->longitude, out->altitude_m,
             out->speed_mps, out->heading_deg, out->serial_prefix);

    return true;
}

/* ── Public API ─────────────────────────────────────────────────────────── */

bool dji_parse_vendor_ie(const uint8_t *payload, size_t len, dji_drone_id_data_t *out)
{
    if (payload == NULL || out == NULL) return false;

    /* Need at least 3 OUI bytes + MIN_PAYLOAD_LENGTH data bytes */
    if (len < 3 + DJI_MIN_PAYLOAD_LENGTH) return false;

    /* Verify DJI OUI */
    if (payload[0] != DJI_OUI_0 || payload[1] != DJI_OUI_1 || payload[2] != DJI_OUI_2) {
        return false;
    }

    /* Strip OUI and parse the remaining payload */
    return parse_payload(&payload[3], len - 3, out);
}

bool dji_parse_beacon_frame(const uint8_t *frame, size_t len, dji_drone_id_data_t *out)
{
    if (frame == NULL || out == NULL) return false;

    /*
     * 802.11 beacon frame layout:
     *   - 24 bytes: MAC header (frame control, duration, addr1-3, seq ctrl)
     *   - 12 bytes: Fixed parameters (timestamp, beacon interval, capability)
     *   - Variable: Tagged parameters (IEs)
     *
     * We start scanning tagged parameters at offset 36.
     */
    const size_t TAGGED_PARAMS_OFFSET = 24 + 12; /* MAC header + fixed fields */

    if (len < TAGGED_PARAMS_OFFSET + 2) return false;

    size_t pos = TAGGED_PARAMS_OFFSET;

    while (pos + 2 <= len) {
        uint8_t tag_id = frame[pos];
        uint8_t tag_len = frame[pos + 1];
        size_t ie_start = pos + 2;

        /* Bounds check */
        if (ie_start + tag_len > len) break;

        if (tag_id == IE_VENDOR_SPECIFIC && tag_len >= 3 + DJI_MIN_PAYLOAD_LENGTH) {
            /* Check DJI OUI at the start of the IE payload */
            if (frame[ie_start] == DJI_OUI_0 &&
                frame[ie_start + 1] == DJI_OUI_1 &&
                frame[ie_start + 2] == DJI_OUI_2) {

                /* Strip OUI and parse */
                const uint8_t *data = &frame[ie_start + 3];
                size_t data_len = tag_len - 3;
                if (parse_payload(data, data_len, out)) {
                    return true;
                }
            }
        }

        pos = ie_start + tag_len;
    }

    return false;
}

File: android/app/src/main/java/com/friendorfoe/detection/DjiDroneIdParser.kt
package com.friendorfoe.detection

import android.net.wifi.ScanResult
import android.os.Build
import android.util.Log

/**
 * Parser for DJI DroneID vendor-specific Information Elements (IEs) found in
 * WiFi beacon frames.
 *
 * DJI drones embed flight telemetry in vendor-specific IEs using OUI 26:37:12.
 * The IE payload contains GPS coordinates, altitude, speed, serial number, and
 * the operator's home point. This provides full position data from a WiFi scan
 * — no Remote ID broadcast required.
 *
 * IE format (after 3-byte OUI):
 *   Byte 0:    Version/type indicator
 *   Bytes 1-4: Serial number (first 4 chars, ASCII) — full serial in later block
 *   Bytes 5-8: Drone longitude (int32, degrees × 1e7)
 *   Bytes 9-12: Drone latitude (int32, degrees × 1e7)
 *   Bytes 13-14: Altitude (int16, meters relative to takeoff point)
 *   Bytes 15-16: Height above ground (int16, decimeters)
 *   Bytes 17-18: Speed (uint16, cm/s)
 *   Bytes 19-20: Heading (int16, degrees × 100, 0 = North)
 *   Bytes 21-24: Home longitude (int32, degrees × 1e7)
 *   Bytes 25-28: Home latitude (int32, degrees × 1e7)
 *
 * Reference: Department of Defense CUAS DroneID analysis, open-source
 * implementations (e.g., kismet, opendroneid-dissector).
 *
 * Requires API 33+ for [ScanResult.getInformationElements].
 */
object DjiDroneIdParser {

    private const val TAG = "DjiDroneIdParser"

    /** DJI's vendor-specific OUI for DroneID IEs */
    private val DJI_OUI = byteArrayOf(0x26, 0x37, 0x12)

    /** Vendor-specific IE element ID per IEEE 802.11 */
    private const val IE_VENDOR_SPECIFIC = 221

    /** Minimum payload length after OUI to contain position data */
    private const val MIN_PAYLOAD_LENGTH = 29

    /**
     * Attempt to parse DJI DroneID data from a WiFi scan result.
     *
     * @param scanResult A WiFi scan result to inspect for vendor IEs
     * @return Parsed [DjiDroneIdData] if a valid DroneID IE is found, null otherwise
     */
    fun parse(scanResult: ScanResult): DjiDroneIdData? {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            return null // getInformationElements() requires API 33
        }

        return try {
            parseApi33(scanResult)
        } catch (e: Exception) {
            Log.w(TAG, "Failed to parse DroneID IE: ${e.message}")
            null
        }
    }

    @Suppress("NewApi")
    private fun parseApi33(scanResult: ScanResult): DjiDroneIdData? {
        val ies = scanResult.informationElements ?: return null

        for (ie in ies) {
            if (ie.id != IE_VENDOR_SPECIFIC) continue

            val bytes = ie.bytes ?: continue
            val payload = ByteArray(bytes.remaining())
            bytes.get(payload)
            bytes.rewind()

            // Check for DJI OUI at the start of the payload
            if (payload.size < 3) continue
            if (payload[0] != DJI_OUI[0] || payload[1] != DJI_OUI[1] || payload[2] != DJI_OUI[2]) continue

            // Strip OUI prefix
            val data = payload.copyOfRange(3, payload.size)
            if (data.size < MIN_PAYLOAD_LENGTH) {
                Log.d(TAG, "DroneID IE too short: ${data.size} bytes (need $MIN_PAYLOAD_LENGTH)")
                continue
            }

            return parsePayload(data)
        }

        return null
    }

    /**
     * Parse the DroneID payload bytes (after OUI) into structured data.
     */
    private fun parsePayload(data: ByteArray): DjiDroneIdData? {
        val version = data[0].toInt() and 0xFF

        // Drone position: degrees × 1e7 as signed 32-bit integer
        val droneLon = readInt32LE(data, 5) / 1e7
        val droneLat = readInt32LE(data, 9) / 1e7

        // Sanity check coordinates
        if (droneLat < -90 || droneLat > 90 || droneLon < -180 || droneLon > 180) {
            Log.d(TAG, "Invalid coordinates in DroneID: lat=$droneLat, lon=$droneLon")
            return null
        }
        if (droneLat == 0.0 && droneLon == 0.0) {
            Log.d(TAG, "Zero coordinates in DroneID — GPS not acquired")
            return null
        }

        // Altitude relative to takeoff (int16, meters)
        val altitudeRelative = readInt16LE(data, 13).toDouble()

        // Height above ground (int16, decimeters → meters)
        val heightAboveGround = readInt16LE(data, 15) / 10.0

        // Speed (uint16, cm/s → m/s)
        val speedCmS = readUInt16LE(data, 17)
        val speedMps = speedCmS / 100.0f

        // Heading (int16, degrees × 100 → degrees)
        val headingRaw = readInt16LE(data, 19)
        val headingDeg = headingRaw / 100.0f

        // Home point / operator position
        val homeLon = readInt32LE(data, 21) / 1e7
        val homeLat = readInt32LE(data, 25) / 1e7

        // Try to extract serial number (ASCII bytes starting at offset 1)
        val serial = try {
            val serialBytes = data.copyOfRange(1, 5)
            String(serialBytes, Charsets.US_ASCII).trim('\u0000')
        } catch (_: Exception) {
            null
        }

        Log.d(TAG, "Parsed DroneID: lat=$droneLat, lon=$droneLon, alt=${altitudeRelative}m, " +
                "speed=${speedMps}m/s, heading=${headingDeg}°, serial=$serial")

        return DjiDroneIdData(
            latitude = droneLat,
            longitude = droneLon,
            altitudeMeters = altitudeRelative,
            heightAboveGroundMeters = heightAboveGround,
            speedMps = speedMps,
            headingDegrees = headingDeg,
            homeLatitude = if (homeLat != 0.0) homeLat else null,
            homeLongitude = if (homeLon != 0.0) homeLon else null,
            serialPrefix = serial,
            version = version
        )
    }

    /** Read a signed 32-bit little-endian integer at the given offset. */
    private fun readInt32LE(data: ByteArray, offset: Int): Int {
        if (offset + 3 >= data.size) return 0
        return (data[offset].toInt() and 0xFF) or
                ((data[offset + 1].toInt() and 0xFF) shl 8) or
                ((data[offset + 2].toInt() and 0xFF) shl 16) or
                ((data[offset + 3].toInt() and 0xFF) shl 24)
    }

    /** Read a signed 16-bit little-endian integer at the given offset. */
    private fun readInt16LE(data: ByteArray, offset: Int): Int {
        if (offset + 1 >= data.size) return 0
        val value = (data[offset].toInt() and 0xFF) or
                ((data[offset + 1].toInt() and 0xFF) shl 8)
        // Sign extend from 16-bit
        return if (value >= 0x8000) value - 0x10000 else value
    }

    /** Read an unsigned 16-bit little-endian integer at the given offset. */
    private fun readUInt16LE(data: ByteArray, offset: Int): Int {
        if (offset + 1 >= data.size) return 0
        return (data[offset].toInt() and 0xFF) or
                ((data[offset + 1].toInt() and 0xFF) shl 8)
    }
}

/**
 * Parsed DJI DroneID telemetry from a WiFi beacon vendor-specific IE.
 *
 * @property latitude Drone latitude (WGS84 degrees)
 * @property longitude Drone longitude (WGS84 degrees)
 * @property altitudeMeters Altitude relative to takeoff point (meters)
 * @property heightAboveGroundMeters Height above ground (meters)
 * @property speedMps Ground speed (m/s)
 * @property headingDegrees Heading (0-360, true north)
 * @property homeLatitude Operator/home point latitude, null if not available
 * @property homeLongitude Operator/home point longitude, null if not available
 * @property serialPrefix First 4 characters of serial number, null if not readable
 * @property version DroneID protocol version byte
 */
data class DjiDroneIdData(
    val latitude: Double,
    val longitude: Double,
    val altitudeMeters: Double,
    val heightAboveGroundMeters: Double,
    val speedMps: Float,
    val headingDegrees: Float,
    val homeLatitude: Double?,
    val homeLongitude: Double?,
    val serialPrefix: String?,
    val version: Int
)
